Add Fill & Outline styles, width settings & new typography controls for Product(Add to cart) button block

imanish003 opened the pull request:
<!-- Start by describing the changes made in this Pull Request and the reason for such changes. -->

This PR is part of #8085 & adds following controls to product(Add to cart) button in Inspector controls.
#### Styles: Fill & Outline
Here is a quick demo video:

https://user-images.githubusercontent.com/16707866/226090143-417c53ad-8de5-4ed4-81c9-ec9628e64471.mov

#### Width Settings
The button block now has a width selector, which allows the user to set the button to 25%, 50%, 75%, or 100% of the parent container. By default, a button's width is determined by the size of its content.

You can find this setting in the button block's sidebar.
![image](https://user-images.githubusercontent.com/16707866/226090241-a2976763-611d-4683-96c3-f6f911c32842.png)

#### Typography controls

Now following typography controls are also available:
- Font Family
- Appearance
- Line height
- Letter spacing
- Decoration
- Letter case


<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

- Create a new post & add `Products block` to it.
- In products block, you can find `Add to cart button` as shown in screenshot below
<img width="343" alt="image" src="https://user-images.githubusercontent.com/16707866/226090398-8479abc2-c69d-416c-b5c2-c444b096a200.png">
- In the sidebar, you will see all newly added controls
<img width="277" alt="image" src="https://user-images.githubusercontent.com/16707866/226090467-d0f7457f-fd56-4dba-a5bc-607ae179f9e5.png">

- Following newly added controls should work as expected:
   - Styles: Fill & Outline ( Feature plugin )
   - Width Settings ( Feature Plugin +  WooCommerce Core )
   - Typography controls ( Feature plugin )

- Also, `block.tsx` file is imported by Cross-sells products [here](https://github.com/woocommerce/woocommerce-blocks/blob/192bdde6c4812dbe626a6444a37a6897d159f2d1/assets/js/blocks/cart/cart-cross-sells-product-list/cart-cross-sells-product.tsx#L18-L19). So make sure that Add to cart button still works fine for Cross-sell products. You can find more about cross-sell products [here](https://woocommerce.com/document/related-products-up-sells-and-cross-sells/).

**Known issues**
- Although all of new controls should also work for Add to cart button in `All products` block, but due to already reported issue [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930), these won't work for now. IMO, This issue should be fixed when we will work on [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930) issue. What do you think?

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [X] Feature plugin
* [ ] Experimental

### Changelog

> Add Fill & Outline styles, width settings & new typography controls for Product(Add to cart) button block.

imanish003 committed:
Add Fill & Outline styles to Product button

This commit adds the Fill & Outline styles to the Product button block.
You can see new Styles section in Inspector Controls where you will be able to change between Fill & Outline styles.

imanish003 committed:
Add width setting to product button

The button block now has a width selector, which allows the user to
set the button to 25%, 50%, 75%, or 100% of the parent container.
By default, a button's width is determined by the size of its content.

You can find this setting in the button block's sidebar.

imanish003 committed:
Enable all typography controls

Now following typography controls are also available:
- Font Family
- Appearance
- Line height
- Letter spacing
- Decoration
- Letter case

imanish003 committed:
Fix width settings not working

imanish003 committed:
Fix align doesn't work with Width settings

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8085-add-to-cart-button-functionalities

imanish003 committed:
Merge branch 'trunk' into add/8085-add-to-cart-button-functionalities

imanish003 committed:
Fix width setting not working in Single product block

imanish003 commented:
Hi @Aljullu, I requested a review from you too, because following will require your opinion too:

> Known issues
Although all of new controls should also work for Add to cart button in All products block, but due to already reported issue https://github.com/woocommerce/woocommerce-blocks/issues/7930, these won't work for now. IMO, This issue should be fixed when we will work on https://github.com/woocommerce/woocommerce-blocks/issues/7930 issue. What do you think?

Please find more details in PR description 🙏🏻

kmanijak commented:
Great work @imanish003!

I tested the solution and I found these two discrepancies:

#### Default button (without outline) is shorter on the frontend.
In Editor default and outlined button has the same height. Outlined button in the frontend as well. But default button on the frontend is shorter (smaller padding)

https://user-images.githubusercontent.com/20098064/226564148-3a7f753c-5bd1-4dc1-97f2-14c33e9f048b.mov

#### 100% width button that is technically an anchor exceeds the container (frontend)
"Select Options" or "View products" for variable and grouped products don't respect the container max width
<img width="429" alt="image" src="https://user-images.githubusercontent.com/20098064/226563825-2a7dbc98-a16c-4e60-b278-b11c92a0ce8f.png">

Also, for small width (25%) they are wider

<img width="399" alt="image" src="https://user-images.githubusercontent.com/20098064/226565271-6c595486-eb15-4ce4-aa14-21ba03861a3b.png">

**EDIT: The issue with width of anchor occurs only in Product Catalog, but not in a post**

---

> Although all of new controls should also work for Add to cart button in All products block, but due to already reported issue https://github.com/woocommerce/woocommerce-blocks/issues/7930, these won't work for now. IMO, This issue should be fixed when we will work on https://github.com/woocommerce/woocommerce-blocks/issues/7930 issue. What do you think?

My opinion is we skip that and focus on Products block, so I agree with you ☝️

kmanijak commented on src/BlockTypes/ProductButton.php:
```diff
@@ -89,8 +89,10 @@ protected function render( $attributes, $content, $block ) {
+			$styles_and_classes            = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
 			$text_align_styles_and_classes = StyleAttributesUtils::get_text_align_class_and_style( $attributes );
+			$classname                     = isset( $attributes['className'] ) ? $attributes['className'] : '';
+			$custom_width_classes          = isset( $attributes['width'] ) ? 'has-custom-width wp-block-button__width-' . $attributes['width'] : '';
```
What do you think about extending the `get_classes_and_styles_by_attributes` method with the `custom_width_classes` so it's there by default?

I did that for `text-align` attribute in this PR:
https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-bf7a423d751aba56975f9ca86523d396e462e59edf9d9d572109cad8617356c8

Or at least move the CSS class extraction to the `StyleAttributesUtils` class so it's shared and available for other blocks as well?

imanish003 replied:
Hi Karol! Thanks for pointing this out 🙌🏻

I am not 100% sure about this. Because this is a custom implementation for Add to cart button for now. You can see the code [here](https://github.com/woocommerce/woocommerce-blocks/blob/60c23add44c8bd2e2d17ebbf8f6f402fc94354ff/assets/js/atomic/blocks/product-elements/button/edit.tsx#L28-L72). Therefore, this functionality isn't shared with any other block for now.

But in future, If we will abstract this into a reusable block, then I agree with you that we will need to extend `get_classes_and_styles_by_attributes`.

What do you think?

kmanijak replied:
Hmm.. 🤔 Makes sense to me, thanks for explanation!

Aljullu commented:
> - Although all of new controls should also work for Add to cart button in `All products` block, but due to already reported issue [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930), these won't work for now.

That's weird, I'm no longer able to reproduce #7930 (I tested changing the color of the Add to Cart button):

Editor | Frontend
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/226619758-4b6e71f7-e20d-4419-9314-006769a09768.png) | ![imatge](https://user-images.githubusercontent.com/3616980/226619809-5588e174-8b4a-40d4-9165-88d0f3b3bf5e.png)

> IMO, This issue should be fixed when we will work on [#7930](https://github.com/woocommerce/woocommerce-blocks/issues/7930) issue. What do you think?

I don't think I'm entitled to make a decision on this. I agree that the priority is the Products block, but at the same time the All Products block is widely used so I'm not sure about having an experience that feels broken. Tagging @danielwrobert for his thoughts on this.

**tl;dr** when adding support for new styles to the _Products_ block product element blocks, they are exposed to the _All Products_ block product element blocks as well. To make these new styling options work in product element blocks inside _All Products_ block, extra work is required.

Options we have:

1. Don't fix it, as the priority right now is the Products block, we shouldn't spend time in the All Products block.
2. Fix it, as the All Products block is widely used and it hasn't been officially deprecated yet.
3. Somewhere in between: time-box some hours to see if it's easy to fix it in the All Products block. If it's easy, let's fix it, otherwise, let's leave it as-is and continue the work with the Products block.

(I lean towards option 3, but that depends on whether it's worth continuing spending time on the All Products block)

danielwrobert commented:
@Aljullu thank you for the ping and for the TL;DR - that's super helpful! 🍻

I agree with option 3. While we don’t want to leave the experience broken for folks who are currently using the All Products block, the Products Block is meant to replace it and we don’t want to let the former hinder the progress of the latter. Time-boxing to determine the necessary effort seems like a good balance.

@imanish003 since we already have an open Issue in #7930 it seems like we’re planning to address this either way - it’s just a matter of _when_ – i.e., do we want to hold up the progress of this PR or, if it requires a more involved fix, merge it and circle back. Is that correct?

imanish003 committed:
Merge branch 'trunk' into add/8085-add-to-cart-button-functionalities

imanish003 commented:
Hi @Aljullu @danielwrobert, Thank you for sharing your thoughts 🙏🏻

> That's weird, I'm no longer able to reproduce https://github.com/woocommerce/woocommerce-blocks/issues/7930 (I tested changing the color of the Add to Cart button):

I also tested again, and it looks like all the controls are working fine **except `Width setting`**, which will be added in this PR. As this is something which will be introduced in this PR, therefore it makes sense that this should be fixed in this PR itself. I will look into it 🙂

Also, I think we may close #7930 if we can't reproduce it anymore.

imanish003 committed:
Fix - Default button is shorter on the frontend

In Editor default and outlined button has the same height. Outlined button in the frontend as well. But default button on the frontend is shorter (smaller padding)

More info: https://github.com/woocommerce/woocommerce-blocks/pull/8781#issuecomment-1477527049

imanish003 committed:
Fix - background color not matching when GB is enabled

Aljullu commented:
Thanks!

> Also, I think we may close https://github.com/woocommerce/woocommerce-blocks/issues/7930 if we can't reproduce it anymore.

I went ahead and closed it. :slightly_smiling_face: 

Edit: ups, by mistake I also closed this PR. I have reopened it. :sweat_smile:

imanish003 committed:
Fix - 100% width button exceeds the container

Fix following issues:
- "Select Options" or "View products" for variable and grouped products don't respect the container max width
- Also, for small width (25%) they are wider

More info: https://github.com/woocommerce/woocommerce-blocks/pull/8781#issuecomment-1477527049

imanish003 commented:
Hi, @kmanijak 👋  I have made the requested changes except [this one](https://github.com/woocommerce/woocommerce-blocks/pull/8781#discussion_r1143114994), in which I have replied to your suggestion for further discussion 🙂

Please let me know if you see any issues, except `Width setting` not working in All products block.

imanish003 committed:
Fix: Width setting not working in All Products block

This commit fixes the width setting not working in the All Products block.

imanish003 committed:
Resolve conflicts

imanish003 commented:
@Aljullu @kmanijak I have made changes in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/417e19612db8aed8257d373cec99acb4b7794e4f) to support Width Settings in `All Products` block. Can you please have a look? Because I am not 100% if that's the right way to do it. 🙏🏻

kmanijak commented:
> Hi, @kmanijak 👋 I have made the requested changes except https://github.com/woocommerce/woocommerce-blocks/pull/8781#discussion_r1143114994, in which I have replied to your suggestion for further discussion 🙂
> Please let me know if you see any issues, except Width setting not working in All products block.

> @Aljullu @kmanijak I have made changes in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/417e19612db8aed8257d373cec99acb4b7794e4f) to support Width Settings in All Products block. Can you please have a look? Because I am not 100% if that's the right way to do it. 🙏🏻

I retested and the issues I reported earlier are gone 💪 However, neither width nor outline is applied to the All Products on the frontend (they work fine in Editor).

EDIT:
After more testing:
- Styles are applied correctly in the "Edit Mode" of All Products
- Only Width is applied correctly in the Editor in "Preview Mode" - outline is ignored
- Width is applied on the frontend from previous "Save" (e.g. if I change settings and save three times, what I see in frontend is the result of second save, not the last third one). Outline is ignored

imanish003 committed:
Merge branch 'trunk' into add/8085-add-to-cart-button-functionalities

imanish003 committed:
Keep other classes along with Width setting classes

`block.attributes.className` should be kept along with the width setting classes.
```
className: classnames( block.attributes.className, {
	[ `has-custom-width wp-block-button__width-${ block.attributes?.width }` ]:
		block.attributes?.width,
} )
```

imanish003 commented:
@kmanijak I fixed the issue in [ffa2c46](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/ffa2c460f3384acf98fbcd2b41fcdcca4a5ed363). Thanks for pointing this out. 🙌🏻

kmanijak commented on assets/js/atomic/blocks/product-elements/button/index.ts:
```diff
@@ -32,6 +34,11 @@ const blockConfig: BlockConfiguration = {
 	edit,
+	save,
+	styles: [
+		{ name: 'fill', label: 'Fill', isDefault: true },
+		{ name: 'outline', label: 'Outline' },
```
I think we could translate labels here

imanish003 replied:
Nice catch! I have added translation for labels in [2cd7e66](https://github.com/woocommerce/woocommerce-blocks/pull/8781/commits/2cd7e661158b31ff25e6e9987f511a77ea5e4d95). Can you please have a quick look? 🙂

kmanijak reviewed (approved):
Works like a charm in Products and All Products blocks! 💪 

I left one small comment, so I'm pre-approving ✅ 

One question: you marked PR as a "Feature Plugin" but I don't see the width setting will enter the WC Core if I'm not mistaken. In that case I think it's better to mark "WC Core", what do you think?

imanish003 commented:
Hi @kmanijak, Thanks for pointing this out 🙌🏻

> One question: you marked PR as a "Feature Plugin" but I don't see the width setting will enter the WC Core if I'm not mistaken. In that case I think it's better to mark "WC Core", what do you think?

You are right, As `Width setting` will be included in WC core too so I am checking both `Feature plugin` & `WooCommerce Core` checkbox for `WooCommerce Visibility`. 

Also, I have updated testing step to include it i.e.:
```
Styles: Fill & Outline ( Feature plugin )
Width Settings ( Feature Plugin + WooCommerce Core )
Typography controls ( Feature plugin )
```

imanish003 committed:
Translate styles label in block configuration

imanish003 committed:
Merge branch 'trunk' into add/8085-add-to-cart-button-functionalities
